internal/docstore/docstore.go

```diff
@@ -0,0 +1,216 @@
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package docstore provides a portable implementation of a document store.
+package docstore // import "gocloud.dev/internal/docstore"
+
+import (
+	"context"
+	"strings"
+	"unicode/utf8"
+
+	"gocloud.dev/internal/docstore/driver"
+	"gocloud.dev/internal/gcerr"
+)
+
+// A Document is a set of field-value pairs.
+// It can be represented as a map[string]int or a pointer to a struct.
+// For structs, the exported fields are the document fields.
+type Document = interface{}
+
+// A Collection is a set of documents.
+type Collection struct {
+	driver driver.Collection
+}
+
+// NewCollection is intended for use by provider implementations.
+var NewCollection = newCollection
+
+// newCollection makes a Collection.
+func newCollection(d driver.Collection) *Collection {
+	return &Collection{driver: d}
+}
+
+// Actions returns an ActionList that can be used to perform
+// actions on the collection's documents.
+func (c *Collection) Actions() *ActionList {
+	return &ActionList{coll: c}
+}
+
+// An ActionList is a sequence of actions that affect a single collection.
+type ActionList struct {
+	coll    *Collection
+	actions []*Action
+}
+
+// An Action is a read or write on a single document.
+// Use the methods of ActionList to create and execute Actions.
+type Action struct {
+	kind       driver.ActionKind
+	doc        Document
+	fieldpaths []string // paths to retrieve, for Get
+	mods       Mods     // modifications to make, for Update
+}
+
+func (l *ActionList) add(a *Action) *ActionList {
+	l.actions = append(l.actions, a)
+	return l
+}
+
+// Create adds an action that creates a new document.
+// The document must not already exist; an AlreadyExists error is returned if it
+// does. If the document doesn't have key fields, it will be given key fields with
+// unique values.
+// TODO(jba): treat zero values for struct fields as not present?
+func (l *ActionList) Create(doc Document) *ActionList {
+	return l.add(&Action{kind: driver.Create, doc: doc})
+}
+
+// Replace adds an action that replaces a document.
+// The document must already exist; a NotFound error is returned if it does not.
+func (l *ActionList) Replace(doc Document) *ActionList {
+	return l.add(&Action{kind: driver.Replace, doc: doc})
+}
+
+// Put adds an action that adds or replaces a document.
+// The document may or may not already exist.
+func (l *ActionList) Put(doc Document) *ActionList {
+	return l.add(&Action{kind: driver.Put, doc: doc})
+}
+
+// Delete adds an action that deletes a document.
+// Only the key fields and RevisionField of doc are used.
```

Comment:
This is the first mention of `RevisionField`...?

Reply:
Let's leave it, I'll be adding RevisionField in a subsequent PR.

Reply:
To be clear, my concern is the toil of copying the right text between my original code and these partial PRs. I've already removed most references to RevisionField in separate sentences; here I'd worry that I'd miss this reference.